### DISCUSSION TITLE
apply derived homography matrix by kornia to cv2 perspective warp

### DISCUSSION THREAD

**Discussion by qiangwang57** [kornia/kornia#1064] [Questions] [answered]
Hi there,

I followed the example of homography regression by kornia and obtained a homography matrix. I wanted to apply the matrix for further post-processing with cv2. Unfortunately, cv2 did not give the correct answer with the matrix. I was just wondering if you could shed some light on it?

Cheers, Chiang

**Comment by shijianjian** [comment]
Hi, can you share a minimum example to reproduce?

**Comment by qiangwang57** [comment]
Hi,

A homography matrix is generated from https://github.com/kornia/kornia/blob/master/examples/homography_regression/main.py:
```
array([[ 1.1086216 , -0.00981381, -0.03928841],
       [-0.05483853,  1.031984  , -0.0259179 ],
       [ 0.12437177,  0.12407163,  0.91941243]], dtype=float32)
```

And the cv2 code is as below:
```
fh_image_file = xxxxx
fh_image = cv2.imread(fh_image_file, cv2.IMREAD_GRAYSCALE)
fh_image = cv2.resize(fh_image, (256, 256))
threshold, _ = cv2.threshold(fh_image, 0, 255, cv2.THRESH_BINARY+cv2.THRESH_OTSU)
fh_image[fh_image < threshold] = 0

h_image_file = xxxxx
h_image = cv2.imread(h_image_file, cv2.IMREAD_GRAYSCALE)
h_image = cv2.resize(h_image, (256, 256))
h_image = cv2.bitwise_not(h_image)
threshold, _ = cv2.threshold(h_image, 0, 255, cv2.THRESH_BINARY+cv2.THRESH_OTSU)
h_image[h_image < threshold] = 0

fh_image_warped = cv2.warpPerspective(fh_image, homo_matrix, dsize=(256, 256))

blending = fh_image_warped*0.6 + h_image*0.4
plt.subplot(221)
plt.imshow(fh_image, cmap="gray")
plt.subplot(222)
plt.imshow(h_image, cmap="gray")
plt.subplot(223)
plt.imshow(fh_image_warped, cmap="gray")
plt.subplot(224)
plt.imshow(blending, cmap="gray")
plt.tight_layout()
plt.show()
```

Any tips would be appreciated!

Cheers, Chiang

**Comment by edgarriba** [comment]
@qiangwang57 is `kornia.geometry.warp_perspective` giving you the correct result ? what's the visual output you get ?

**Comment by qiangwang57** [comment]
> @qiangwang57 is `kornia.geometry.warp_perspective` giving you the correct result ? what's the visual output you get ?

I used HomographyWarper for the regression, and the result looks great. One thing here is that with the same matrix, warp_perspective did not give the same result. Not sure what is the trick in it... But seems cv2 and warp_perspective give closer result?
![Figure_1](https://user-images.githubusercontent.com/49925149/120651606-e1e50600-c476-11eb-9643-3d218b4cefd4.png)

**Comment by edgarriba** [comment]
What's the call you do for kornia warp_perspective ?

**Comment by qiangwang57** [comment]
@edgarriba here is the code for the three:

```
def cv2_warp(moving_image, ref_image, homo_matrix):
    moving_image_warped = cv2.warpPerspective(moving_image, homo_matrix_inv, dsize=(256, 256))
    blending = moving_image_warped * 0.6 + ref_image * 0.4
    return blending


def homographywaper(moving_image, ref_image, homo_matrix):
    moving_image_tensor = image2tensor(moving_image)
    homo_matrix_tensor = torch.Tensor(homo_matrix)
    homo_matrix_tensor = torch.unsqueeze(homo_matrix_tensor, dim=0)
    warper = HomographyWarper(256, 256)
    moving_image_tensor_warped = warper(moving_image_tensor, homo_matrix_tensor)
    moving_image_warped = (tensor2image(moving_image_tensor_warped)*255).astype(np.uint8)
    blending = moving_image_warped * 0.6 + ref_image * 0.4
    return blending


def warp_perspective_test(moving_image, ref_image, homo_matrix):
    moving_image_tensor = image2tensor(moving_image)
    homo_matrix_tensor = torch.Tensor(homo_matrix)
    homo_matrix_tensor = torch.unsqueeze(homo_matrix_tensor, dim=0)
    moving_image_tensor_warped = warp_perspective(moving_image_tensor, homo_matrix_tensor, dsize=(256, 256))
    moving_image_warped = (tensor2image(moving_image_tensor_warped)*255).astype(np.uint8)
    blending = moving_image_warped * 0.6 + ref_image * 0.4
    return blending
```

**Comment by edgarriba** [comment]
Oh I see it now. That’s the thing, you refine the transformation with `HomographyWarper` which for numerical stability we internally use a normalised homography. However, opencv and kornia `warp_perspective` assume homographies in pixel space. In order to achieve same results you need to normalise first the homography - we have a utility function for that. See here: https://github.com/kornia/kornia/blob/e4a54c2d4fd9b8eaa88dec50061e33f7aee5d959/kornia/geometry/transform/homography_warper.py#L301

Besides, this looks quite a cool application what you do. May I ask you what’s your use case and purpose. We would to extend to an image registration api, and this looks like it’s what you want to achieve. /cc @ducha-aiki @lferraz @ethanrublee

**Reply by LoiseauNicolas** [reply]
Hello, Sorry to re-open an old question. But What is the mean of this homography normalization ? And how it work ? 
Thank you

**Reply by ducha-aiki** [reply]
It is simply the question of working in pixel coordinates or normalized. 
Normalized homography assumes that your topleft pixel is (-1,-1) and bottomright - (1,1), where as unnormalized homography - topleft is (0,0), topright is, say (640, 480)

**Comment by qiangwang57** [comment]
@edgarriba Thanks for the explanation and sorry that I am a bit new to the area. I just wonder if you could shed more light on how to do the trick. You mentioned the homo matrix needs to be normalised for `warp_perspective`, but the utility function is to normalise the pixel space, I assume? I inserted `homo_matrix_tensor = normalize_homography(homo_matrix_tensor, (256, 256), (256, 256))` to kornia's 'warp_perspective', but no luck. 

The application is to align fluorescence microscopic images (the first) with pathology images (the second). Pathology images are the gold standard for fluorescence images, and hence to interrogate fluorescence images at pixel level, the registration is required. And this is fundamental for all other consequent purposes, e.g., segmentation. Regression would be effective in the case presented, but more advanced methods are desirable as there may involve non-rigid registration. Please do include more registration methods in kornia! If needed, I could give a hand on it.

**Comment by ducha-aiki** [comment, marked as answer]
There are 2 ways, I can recommend you for now (until we provide smth out of the box).
First, is to simply use ` kornia.geometry.homography_warp` for warping instead of OpenCV. 
`warped2 = kornia.geometry.homography_warp(img_src, homo_matrix, img_src.shape[-2:])`

Second, use the following function (we will add it to kornia in a couple of days)

```
def denormalize_homography(norm_homo, height, width):
    normal_transform_pixel = kornia.normal_transform_pixel(height, width, device=norm_homo.device)
    denormed_homo = torch.inverse(normal_transform_pixel) @ norm_homo @ normal_transform_pixel
    return denormed_homo
```
Than 
`homo_matrix_tensor = denormalize_homography(homo_matrix_tensor, (256, 256))` would work for kornia.warp_perspective, or its numpy version -- for cv2.warpPerspective

**Comment by qiangwang57** [comment]
> There are 2 ways, I can recommend you for now (until we provide smth out of the box).
> First, is to simply use ` kornia.geometry.homography_warp` for warping instead of OpenCV.
> `warped2 = kornia.geometry.homography_warp(img_src, homo_matrix, img_src.shape[-2:])`
> 
> Second, use the following function (we will add it to kornia in a couple of days)
> 
> ```
> def denormalize_homography(norm_homo, height, width):
>     normal_transform_pixel = kornia.normal_transform_pixel(height, width, device=norm_homo.device)
>     denormed_homo = torch.inverse(normal_transform_pixel) @ norm_homo @ normal_transform_pixel
>     return denormed_homo
> ```
> 
> Than
> `homo_matrix_tensor = denormalize_homography(homo_matrix_tensor, (256, 256))` would work for kornia.warp_perspective, or its numpy version -- for cv2.warpPerspective

Thanks @ducha-aiki ! The first recommendation just works perfectly! However, the second one seems not, very close though.
The code is as below:

```
def cv2_warp(moving_image, ref_image, homo_matrix):
    moving_image_warped = cv2.warpPerspective(moving_image, homo_matrix, dsize=(256, 256))
    blending = moving_image_warped * 0.6 + ref_image * 0.4
    return blending

def warp_perspective_test(moving_image, ref_image, homo_matrix):
    moving_image_tensor = image2tensor(moving_image)
    moving_image_tensor_warped = warp_perspective(moving_image_tensor, homo_matrix, dsize=(256, 256))
    moving_image_warped = (tensor2image(moving_image_tensor_warped)*255).astype(np.uint8)
    blending = moving_image_warped * 0.6 + ref_image * 0.4
    return blending

homo_matrix_tensor = torch.Tensor(homo_matrix)
homo_matrix_tensor = torch.unsqueeze(homo_matrix_tensor, dim=0)
homo_matrix_tensor_denorm = denormalize_homography(homo_matrix_tensor, 256, 256)
blending_cv2 = cv2_warp(fh_image, h_image, homo_matrix_tensor_denorm.squeeze(0).detach().cpu().numpy())
blending_hw = warp_perspective_test(fh_image, h_image, homo_matrix_tensor_denorm)
## plotting....
```

Here is the result.
![Figure_1](https://user-images.githubusercontent.com/49925149/120828225-2db9ad00-c554-11eb-979a-2605d1754f18.png)

Did I miss something in the code?

Cheers, Chiang

**Comment by ducha-aiki** [comment]
You also may need to invert your matrix, I may have missed that part. One function is expecting src_to_dst transformation and another - other way round.

**Comment by qiangwang57** [comment]
Brilliant! Everything works just perfectly. Thank you all for the help!

Cheers, Chiang

**Reply by jc138691** [reply]
Hi, is there some notes on how to use denormalize_homography. I was expecting to have identical results with homo and denormalize_homography:

img2 = KG.homography_warp(
            patch_src=img1, src_homo_dst=homo, dsize=dsize, padding_mode='border')

de_homo = kornia.geometry.conversions.denormalize_homography(
            homo, dsize_src=dsize, dsize_dst=dsize)

img2_c = KG.homography_warp(
            patch_src=img1,
            src_homo_dst=de_homo,
            normalized_homography=False,
            dsize=dsize, padding_mode='border')

img2 and img2_c should be identical but they are not. Any suggestions?

**Reply by edgarriba** [reply]
could you share the full snippet so that we can reproduce? maybe a colab is enough
